code/HasCanonicalEditLink.php

```diff
@@ -0,0 +1,152 @@
+<?php
+
+namespace SilverStripe\Admin;
+
+use LogicException;
+use SilverStripe\CMS\Controllers\CMSMain;
+use SilverStripe\Control\Controller;
+use SilverStripe\Control\Director;
+use SilverStripe\Forms\FieldList;
+use SilverStripe\Forms\GridField\GridField;
+use SilverStripe\Forms\GridField\GridFieldDetailForm;
+use SilverStripe\ORM\DataObject;
+
+/**
+ * A trait that automatically generates a CMS edit link for DataObjects even if
+ * they are canonically edited in some nested {@link GridField}.
+ * Designed to be used in conjunction with the {@link CMSPreviewable} interface.
+ *
+ * For nested relations (e.g. a DataObject managed in a GridField of another DataObject)
+ * you can apply this trait to both the parent and the child object and the links
+ * will chain down the nested `GridField`s to the root canonical edit owner.
+ *
+ * You must set a canonical_edit_owner config variable which defines the canonical
+ * owner for this class.
+ * e.g. set this to a {@link LeftAndMain} class:
+ * private static string canonical_edit_owner = MyModelAdmin::class;
+ * or to a has_one relation:
+ * private static string canonical_edit_owner = 'Parent';
+ *
+ * Note that the canonical edit owner must implement a getEditLinkForObject() method.
+ *
+ * If the canonical edit owner is a has_one relation, the class on the other end
+ * of the relation must have a CMSEditLink() method.
+ */
+trait HasCanonicalEditLink
```

Comment:
I was out for a bit and may have missed it, but do we have a naming convention for traits?

I'm a proponent of suffixing the name with the type of class/construct it is, e.g. `SomethingTrait`, `PageController`, `FieldExtension` etc. so it's immediately obvious what it is.

```suggestion
trait CanonicalEditLinkTrait
```

Reply:
We do have quite a few traits already - some are suffixed with `Trait` but the majority are not.
I think the fact that you declare it _as_ a trait, and you `use` it, make it abundantly clear it's a trait.

`Extension` and `Controller` are a bit different because those aren't built-in language structure things like a trait (we don't say `SiteTreeClass` after all).

Reply:
Agree we shouldn't suffix `Trait` on this one (even though I'd actually like it, I also like the `Interface` suffix), because most of the code base doesn't suffix it

I don't like the use of the word `Canonical` at all - it really should just be `CMSEditLink`

I had a look at different traits in the codebase to find a common naming convention, there isn't that much in the way of consistency.

The most common pattern I could find was just calling the trait the feature it implemented e.g.
- trait ImageManiputlation
- trait FlushScheduler
- trait FormMessage

The next most consistent, though less common one was the Configurable / Extensible / Injectable - though those are less "concrete" concepts than just the other traits that just bolt a feature on

This makes a couple of obvious candidates
- trait CMSEditLink
- trait CMSEditLinkable (weird IMO)

I think `CMSEditLink` is the most non-horrible of the two

```php
class MyClass
{
   use CMSEditLink;
```
And I get a `CMSEditLink()` method

Reply:
I'll rename the trait to `CMSEditLink`